packages/salesforcedx-utils-vscode/src/services/telemetry.ts

```diff
@@ -383,4 +385,65 @@ export class TelemetryService {
         .catch(err => console.error(err));
     }
   }
+
+  private getRandomUserId = (): string => randomBytes(20).toString('hex');
+
+  private buildCliTelemetryCommand(): Command {
+    return new SfCommandBuilder()
+      .withArg('telemetry')
+      .withJson()
+      .build();
+  }
+
+  private async executeCliTelemetry(
+    command: Command
+  ): Promise<string> {
+    const workspacepath = workspaceUtils.getRootWorkspacePath();
+    const execution = new CliCommandExecutor(command, { cwd: workspacepath }
+    ).execute();
+    const cmdOutput = new CommandOutput();
+    const result = cmdOutput.getCmdResult(execution);
+    return result;
+  }
+
+  public async getUserId(): Promise<string> {
+    // Defining UserId in globalState and using the same in appInsights reporter.
+    // Assigns cliId to UserId when it's undefined in global state.
+    // cliId is undefined when cli-telemetry variable disable-telemetry is true.
+    // Relying on machineId as backup option.
+    let globalStateUserId;
+
+    try {
+      globalStateUserId = this.extensionContext?.globalState.get<string | undefined>('UserId');
+      if (globalStateUserId === undefined) {
+        const getCliTelemetryData = await this.executeCliTelemetry(
+          this.buildCliTelemetryCommand()
+        );
+        const cmdResult = JSON.parse(getCliTelemetryData) as { result?: { cliId: string } };
+        let cliUserId = cmdResult?.result?.cliId;
+
+        // Generates a random UserId when cliId is undefined.
+        if (cliUserId === undefined) {
+          cliUserId = this.getRandomUserId();
+        }
+        await this.extensionContext?.globalState.update('UserId', cliUserId);
+        globalStateUserId = cliUserId;
+      }
+    } catch (error) {
+      console.log(`Error: ${error} occurred in retrieving cliId, generating user-id ..`);
+      globalStateUserId = this.extensionContext?.globalState.get<string | undefined>('UserId');
+
+      // Assigns a random value to UserId when it's undefined and sf telemetry cmd throws an error.
+      if (globalStateUserId === undefined) {
+        globalStateUserId = this.getRandomUserId();
+        await this.extensionContext?.globalState.update('UserId', globalStateUserId);
+      }
+    } finally {
+      if (globalStateUserId === undefined) {
+        globalStateUserId = env.machineId;
+      }
+    }
+
+    return globalStateUserId;
+  }
```

Comment:
Just a suggestion. Using finally to set a value is unreliable given the finally block is run when garbage collection recovers the memory. I also don't see a need to use machineId at all.
``` suggestion
  public async getUserId(): Promise<string> {
    // Defining UserId in globalState and using the same in appInsights reporter.
    // Assigns cliId to UserId when it's undefined in global state.
    // cliId is undefined when cli-telemetry variable disable-telemetry is true.
    let globalStateUserId = this.extensionContext?.globalState.get<
      string | undefined
    >('UserId');

    if (globalStateUserId) {
      return globalStateUserId;
    }

    globalStateUserId = await this.executeCliTelemetry(
      this.buildCliTelemetryCommand()
    )
      .then((getCliTelemetryData): string => {
        const cmdResult = JSON.parse(getCliTelemetryData) as {
          result?: { cliId: string };
        };
        return cmdResult?.result?.cliId ?? this.getRandomUserId();
      })
      .catch(error => {
        console.log(
          `Error: ${error} occurred in retrieving cliId, generating user-id ..`
        );
        return this.getRandomUserId();
      });

    await this.extensionContext?.globalState.update(
      'UserId',
      globalStateUserId
    );

    return globalStateUserId;
  }
```

Reply:
good suggestion 👍